### DISCUSSION TITLE
Understanding persistence

### DISCUSSION THREAD

**Discussion by gcheron** [milvus-io/milvus#29662] [Q&A and General discussion]
Hi!

I am trying to setup a **standalone** Milvus on **EKS** with **helm**.
My problem is that I do not exactly get what must be persisted.

First, I have disabled `minio` to store objects on my own S3 bucket.

Also, for `standalone`, I pass a PVC called `xxx` that I created myself externally:

```
persistence:
      persistentVolumeClaim:
        existingClaim: xxx
```

But I cannot find similar config for `etcd`, I cannot find a way to pass an existing PVC.
Is it because we do not need to persist `etcd` data ?

My goal is to be able to recover my data after e.g. scaling to zero or even full `helm uninstall`.

1. What will happen to `etcd` if I reinstall with my current persisting data (S3 + standalone PVC)? Will it recover and recompute its necessary data or should I persist `etcd` data? If yes, how?
2. More generally, what data (and how) should be persisted?
3. Last question: including in **Cluster** mode, when creating my external PVCs, can I use EBS everywhere or is there  configurations where volumes must be mounted at multiple locations (then I would have to create EFS disks)?

Many thanks!

**Comment by yhmo** [comment]
Milvus stores system metadata in Etcd, and stores data files in S3
- system metadata: collections schema, RBAC user/password, the S3 path of each collection/partition/segment, etc.
- data files: column-based files to stores data of fields for each collection, index files to store index after you call create_index(), bloom-filter files to quickly find entity by id.

The metada in Etcd must be strongly consist with the data files in S3, otherwise milvus will run into errors.

Theoretically, manually copy the etcd/s3 data folders(stop milvus server first) can backup the entire data, refer to this discussion: https://github.com/milvus-io/milvus/discussions/29277

However we don't recommend manually backup in this way. The official tool [milvus-backup](https://github.com/zilliztech/milvus-backup) is recommended, the tool can backup collections and restore to another milvus cluster. Another tool [milvus-cdc](https://github.com/zilliztech/milvus-cdc) is in development for capturing data changes between milvus clusters.

If you insist to find the PVC of the etcd pod, let me call someone to help.

**Reply by gcheron** [reply]
Thanks for the clarification about system metadata.
In standalone mode, what is saved in the standalone persistence?

Regarding to backup, I was wondering (in https://github.com/milvus-io/milvus/discussions/29662#discussioncomment-8011467) if instead of backup: can't I reuse a volume that wasn't deleted during uninstallation?

**Reply by yhmo** [reply]
Standalone is using a local minio docker container as storage.
Standalone is deployed by docker-compose, all the data path are mapped to outside(local path), yes you can reuse the volumes to start another milvus standalone, but not able to apply to a cluster.
In the docker-component.yaml of standalone we can see:
```
services:
  etcd:
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/etcd:/etcd
  minio:
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/minio:/minio_data
  standalone:
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/milvus:/var/lib/milvus
```

**Comment by wyfeng001** [comment]
1. You can refer to this document for instructions on how to deploy Milvus on AWS EKS. [https://milvus.io/docs/eks.md](https://milvus.io/docs/eks.md)
2. We used the native[ Etcd Chart ](https://github.com/bitnami/charts/tree/main/bitnami/etcd)with only a few initialization modifications. Not being able to support PVC is not Milvus' behavior
3. From the design of K8S, PVC is related to Pod, but not to PV.  etcd is a statefulset and it uses its own volumeClaimTemplate.
4. I think you can try using backup & restore methods for etcd.  See this [https://docs.bitnami.com/kubernetes/infrastructure/etcd/administration/backup-restore/](https://docs.bitnami.com/kubernetes/infrastructure/etcd/administration/backup-restore/)

**Reply by gcheron** [reply]
Thanks! Thank you for the EKS deployment link too, but I have already went through and I cannot find information about persistence.
Thanks for the clarification about etcd, I get it much better now. So will I be able to recover from the etcd is a StatefulSets if:
- I scale to 0 then 1?
- helm uninstall then helm install?

--> If I understand what you say, here I must backup with etcd? Can't I reuse a volume that wasn't deleted during uninstallation?

**Reply by wyfeng001** [reply]
- Scaling sts would use the existing PVCs but you cannot delete the sts (or helm uninstall). That will recreate a new sts.
- It seems that bitami/etcd could use a PV label selector.  Can have a try！
Remove PVC (BE CAREFUL, note the PV RECLAIM POLICY, some cloud providers may remove PVs at the same time as PVCs. are removed)
And then set this selector to the PV's label 
![image](https://github.com/milvus-io/milvus/assets/28752231/40848818-ef66-4553-8d06-d22542a0f467)

If this is unsuccessful, there may be only a backup recovery method.  Take a closer look at how Etcd Chart does it.

**Comment by wyfeng001** [comment]
If your helm release name is the same as before, you can reuse the saved pvc.
